src/stream/src/executor/source/source_executor.rs

```diff
@@ -156,82 +164,96 @@ impl<S: StateStore> SourceExecutor<S> {
             )
             .inc();
         if let Some(target_splits) = split_assignment.get(&self.actor_ctx.id).cloned() {
-            if let Some(target_state) = self.update_state_if_changed(Some(target_splits)).await? {
-                tracing::info!(
-                    actor_id = self.actor_ctx.id,
-                    state = ?target_state,
-                    "apply split change"
-                );
-
-                self.replace_stream_reader_with_target_state(
-                    source_desc,
-                    stream,
-                    target_state.clone(),
-                )
-                .await?;
-
-                return Ok(Some(target_state));
+            if self
+                .update_state_if_changed(target_splits, should_trim_state)
+                .await?
+            {
+                self.rebuild_stream_reader(source_desc, stream).await?;
             }
         }
 
-        Ok(None)
+        Ok(())
     }
 
-    /// Returns `target_states` if split changed. Otherwise `None`.
-    ///
-    /// Note: `update_state_if_changed` will modify `updated_splits_in_epoch`
+    /// Returns `true` if split changed. Otherwise `false`.
     async fn update_state_if_changed(
         &mut self,
-        state: ConnectorState,
-    ) -> StreamExecutorResult<ConnectorState> {
+        target_splits: Vec<SplitImpl>,
+        should_trim_state: bool,
+    ) -> StreamExecutorResult<bool> {
         let core = self.stream_source_core.as_mut().unwrap();
 
-        let target_splits: HashMap<_, _> = state
-            .unwrap()
+        let target_splits: HashMap<_, _> = target_splits
             .into_iter()
             .map(|split| (split.id(), split))
             .collect();
 
-        let mut target_state: Vec<SplitImpl> = Vec::with_capacity(target_splits.len());
+        let mut target_state: HashMap<SplitId, SplitImpl> =
+            HashMap::with_capacity(target_splits.len());
 
         let mut split_changed = false;
 
         // Checks added splits
-        for (split_id, split) in &target_splits {
-            if let Some(s) = core.updated_splits_in_epoch.get(split_id) {
-                // existing split, no change, clone from cache
-                target_state.push(s.clone())
+        for (split_id, split) in target_splits {
+            if let Some(s) = core.latest_split_info.get(&split_id) {
```

Comment:
This line is a minor change. I think using `latest_split_info` is better than `updated_splits_in_epoch` here.